### ISSUE TITLE
CA1508 false positive in for loop removing items from List

### ISSUE THREAD

**Issue by ReneVogt** [dotnet/roslyn-analyzers#6453]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

Not sure. I reproduced the problem targeting netstandard2.0 and net6.0 with these project settings:

```
 <PropertyGroup>
  <EnableNETAnalyzers>true</EnableNETAnalyzers>
  <AnalysisLevel>latest</AnalysisLevel>
  <AnalysisMode>AllEnabledByDefault</AnalysisMode>
  <LangVersion>latest</LangVersion>
</PropertyGroup>
```

### Describe the bug

In a for loop that compares members of two elements of a `List<>` and then removes an element from that list, the second comparison raises CA1508

### Steps To Reproduce

In this code:

```
  sealed class Data
  {
      public int Value {get;}
      public Data(int Value) 
      {
          Value = value;
      }
  }
 
  static class Test
  {
     static void Filter(List<Data> list)
     {
         for (int i=0; i< list.Count-1; i++)
         {
             if (list[i+1].Value != 0) continue;
             if (list[i].Value == 0) continue; // <-------- CA1508
             list.RemoveAt(i);
         }
    }
  }
```

a CA1508 is raised for the commented line stating that the comparison was always `true`.
 
**Important:** 
1. The order is important. If `list[i]` is compared first CA1508 is not raised
2. The `==` is important. If the second comparision was `list[i].Value != 0` CA1508 is not raised
3. the `RemoveAt()` is important, if omitted, CA1508 is not raised
4. it must be a member access, with only a `List<int>` CA1508 is not raised

### Expected behavior

CA1508 should not be raised, because the comparsion is not always `true`.

### Actual behavior

CA1508 is raised.

## Additional context

In my original code the list is an `ImmutableArray<T>.Builder`.
